engine/src/main/java/org/hibernate/search/backend/impl/DeleteByQuerySupport.java

```diff
@@ -41,221 +29,36 @@ private DeleteByQuerySupport() {
-									value = NumericUtils.doubleToSortableLong( (Double) query.getValue() );
-								}
-								else {
-									value = (Long) query.getValue();
-								}
-								BytesRefBuilder builder = new BytesRefBuilder();
-								NumericUtils.longToPrefixCoded( value, 0, builder );
-								valueAsBytes = builder.get();
-								break;
-							}
-							default:
-								throw new AssertionFailure( "has to be a Numeric Type at this point!" );
-						}
-						return new TermQuery( new Term( query.getFieldName(), valueAsBytes ) );
-					}
-
-				}
-
-			} );
-
-			TO_LUCENE_QUERY_CONVERTER = new ToLuceneQuery[map.size()];
-			for ( Map.Entry<Integer, ToLuceneQuery> entry : map.entrySet() ) {
-				TO_LUCENE_QUERY_CONVERTER[entry.getKey()] = entry.getValue();
-			}
+	public static DeletionQuery fromString(int queryKey, String[] string) {
+		switch ( queryKey ) {
+			case SingularTermDeletionQuery.QUERY_KEY:
+				return SingularTermDeletionQuery.fromString( string );
+			default:
+				throw log.unknownDeletionQueryKeySpecified( queryKey );
 		}
 	}
 
-	public static ToLuceneQuery getToLuceneQuery(int queryKey) {
-		return TO_LUCENE_QUERY_CONVERTER[queryKey];
-	}
-
-	private static final Set<Class<? extends DeletionQuery>> SUPPORTED_TYPES;
-	static {
-		{
-			Map<Integer, Class<? extends DeletionQuery>> map = new HashMap<>();
-
-			map.put( SingularTermQuery.QUERY_KEY, SingularTermQuery.class );
-
-			SUPPORTED_TYPES = Collections.unmodifiableSet( new HashSet<>( map.values() ) );
-		}
-	}
+	private static final Lock SUPPORTED_TYPES_LOCK = new ReentrantLock();
+	private static Set<Class<? extends DeletionQuery>> SUPPORTED_TYPES;
 
 	public static boolean isSupported(Class<? extends DeletionQuery> type) {
-		return SUPPORTED_TYPES.contains( type );
-	}
-
-	private static final StringToQueryMapper[] FROM_STRING;
-	static {
-		{
-			Map<Integer, StringToQueryMapper> map = new HashMap<>();
-
-			map.put( SingularTermQuery.QUERY_KEY, new StringToQueryMapper() {
-
-				@Override
-				public DeletionQuery fromString(String[] string) {
-					if ( string.length != 3 ) {
-						throw new IllegalArgumentException( "for a TermQuery to work there have to be " + "exactly 3 Arguments (type & fieldName & value" );
-					}
-					Type type = Type.valueOf( string[0] );
-					switch ( type ) {
-						case STRING:
-							return new SingularTermQuery( string[1], string[2] );
-						case INT:
-							return new SingularTermQuery( string[1], Integer.parseInt( string[2] ) );
-						case FLOAT:
-							return new SingularTermQuery( string[1], Float.parseFloat( string[2] ) );
-						case LONG:
-							return new SingularTermQuery( string[1], Long.parseLong( string[2] ) );
-						case DOUBLE:
-							return new SingularTermQuery( string[1], Double.parseDouble( string[2] ) );
-						default:
-							throw new AssertionFailure( "wrong Type!" );
-					}
-
+		if ( SUPPORTED_TYPES == null ) {
+			// LAZY init, but don't to it more than once
+			SUPPORTED_TYPES_LOCK.lock();
```

Comment:
You still do the locking thingy here. I'd if/else on the given class and you are done. Esp. since there is no way to plug in another type anyways. Or is there? Also the single user of this method raises an exception in case the type would be "not supported". This all seems still a bit odd to me, at least I don't see how it would work.